src/eko/evolution_operator/__init__.py

```diff
@@ -685,37 +682,46 @@ def a_em(self):
 
     def compute_aem_list(self):
         """Return the list of the couplings for the different values of :math:`a_s`."""
+        ev_op_iterations = self.config["ev_op_iterations"]
```

Comment:
Can you please add a bit more documentation/comments to this function, which is fairly complicated?

Again, maybe open an issue: we may want to reconsider the split between `Operator` and `kernels` ... If I understand this function correctly here you're splitting the evolution integral, which so far was done below (see e.g. [here](https://github.com/NNPDF/eko/blob/45682d7775210f262529d0f639fd1d21a48b5457/src/eko/kernels/singlet.py#L331)), but now you no longer can't because you need to call a function (from `Couplings`). - Actually, why are you using `compute` (internal) and not `a` (public)? this way you never do the matching! - Coming back to the potential issue: now you're sensitive to how you choose the steps, and (by chance) you're currently consistent [here](https://github.com/NNPDF/eko/blob/47d46ce671f78fededb201604fc68e0a35532c06/src/eko/evolution_operator/__init__.py#L706) and [below](https://github.com/NNPDF/eko/blob/47d46ce671f78fededb201604fc68e0a35532c06/src/eko/kernels/non_singlet_qed.py#L254). Instead, we maybe want to compute a list of tuples `[(mu0,as0,a0),(mu1,as1,a1),...]` here (because we have to) and pass that around

Reply:
What this function does is the following:

In the case of running `aem` we need the values of `aem` corresponding to the middle point of the `as` interval, so we need to find $a_{em}(a_s)$. This can be done for the exact evolution (you just write $da_{em}/da_s = \beta_{qed} / \beta_{qcd}$ and you solve numerically) but not for the expanded evolution (or at least I wasn't able to solve the differential equation perturbatively).

For this reason I splitted the $mu^2$ interval in steps and for every step I compute the value of $a_s$ at the borders of the interval (these are now what before was computed inside kernels as `utils.geomspace(a0, a1, 1 + ev_op_iterations)`) and the values of $a_{s,em}$ in the middle point of the $mu$-interval (this is what is going in the computation of the $\gamma$ and $\beta$ in `eko_iterate`)
I know that this is not exactly the same than computing the values of $a_{em}(\frac{a_s^k+a_s^{k+1}}{2}) with $a_s^k$ divided with `utils.geomspace(a0, a1, 1 + ev_op_iterations)` but the distance between the $a_s$ is very small and moreover it is just a different way of splitting the $a_s$ integration interval.

Regarding your questions:

I'm using `compute` and not `a` since `a` always starts from `Qref` while I wanted to evolve from either `q2_to` or `q2_from` (the one that is closer to `Qref`), given that we have already computed the $a_s$ and $a_{em}$ values in these points. In this way the numerical error should be smaller when computing a point that is close to `as1` or `as0`. Then, you are right that I never do the mathcing of `as` but it is not needed since the evolution is computed in one fixed flavor interval at the time (the matching is done when we compute `as0` and `as1`).

Reply:
moved to #224